tools/parallelize/parallelize.cpp

```diff
@@ -0,0 +1,203 @@
+        }
+
+        nextCommandToRun = oldCommandsSize + 1;
+        subprocesses[scheduledOn].schedule(oldCommandsSize);
+        update_display();
+    }
+
+    void wait_all() noexcept {
+        std::unique_lock lck{mtx};
+        cv.wait(lck, [this] { return runningConcurrency.load(std::memory_order_relaxed) == 0; });
+    }
+
+    const std::vector<std::pair<std::wstring, std::optional<execution_result>>>& results() const {
+        assert(runningConcurrency.load(std::memory_order_relaxed) == 0);
+        return commands;
+    }
+
+private:
+    struct entry;
+    std::mutex mtx{};
+    std::condition_variable cv{};
+    std::vector<std::pair<std::wstring, std::optional<execution_result>>> commands{};
+    std::size_t nextCommandToRun{0};
+    std::size_t availableConcurrency{std::thread::hardware_concurrency()};
+    std::atomic<std::size_t> runningConcurrency{0};
+    std::unique_ptr<entry[]> subprocesses;
+
+    struct entry {
+        parallelizer* parent;
+        std::size_t commandRunning{SIZE_MAX}; // guarded by parent->mtx
+        subprocess_executive executive;
+        tp_wait tpWait{callback, this, nullptr};
+
+        static void __stdcall callback(
+            PTP_CALLBACK_INSTANCE, void* thisRaw, PTP_WAIT, TP_WAIT_RESULT waitDisposition) noexcept {
+            assert(waitDisposition == WAIT_OBJECT_0);
+            (void) waitDisposition;
+            const auto this_  = static_cast<entry*>(thisRaw);
+            const auto parent = this_->parent;
+            auto results      = this_->executive.complete();
+            std::lock_guard lck{parent->mtx};
+            parent->commands[this_->commandRunning].second.emplace(std::move(results));
+            if (parent->nextCommandToRun == parent->commands.size()) {
+                this_->commandRunning = SIZE_MAX;
+                if (parent->runningConcurrency.fetch_sub(1, std::memory_order_relaxed) == 1) {
+                    parent->cv.notify_all();
+                }
+            } else {
+                this_->schedule(parent->nextCommandToRun++);
+            }
+
+            parent->update_display();
+        }
+
+        void schedule(std::size_t command) {
+            commandRunning = command;
+            executive.begin_execution(nullptr, parent->commands[command].first.data(), 0, nullptr);
+            tpWait.wait_for(executive.get_wait_handle());
+        }
+    };
+
+    void update_display() noexcept {
+        const auto commandsSize = commands.size();
+        const auto next         = nextCommandToRun;
+        const auto running      = runningConcurrency.load(std::memory_order_relaxed);
+        assert(running <= next);
+        printf("%zu scheduled %zu completed %zu running\n", commandsSize, next - running, running);
+    }
+};
+
+inline void schedule_command(parallelizer& p, const std::wstring_view commandPrefix, const std::wstring& native) {
+    std::wstring toExecute;
+    toExecute.reserve(commandPrefix.size() + native.size() + 1);

```

Comment:
Would putting the `+ 1` in the middle match the sequence of operations below better?